### PR TITLE
add option 'threshold' representing % of different pixels (number between 0 and 100)

### PR DESCRIPTION
**Pull request by nikchosh**
Threshold should be a value between 0 and 100 (% of pixels that are
different). Should take into account that compare.exe sometimes returns normalized numbers of kind 9.9999e-999)

### CODE REVIEW

File: lib/image-diff.js [hunk 1]
```diff
@@ -73,7 +73,7 @@ ImageDiff.createDiff = function (options, cb) {
     // DEV: According to http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=17284
     // DEV: These values are the total square root mean square (RMSE) pixel difference across all pixels and its percentage
     // TODO: This is not very multi-lengual =(
-    var resultInfo = stderr.match(/all: (\d+\.?\d*) \((\d+\.?\d*)\)/);
+    var resultInfo = stderr.match(/all: (\d+\.?\d*) \((.*?)\)/);
```
Comment (twolfson) [inline lib/image-diff.js]: We were intentionally keeping the regexp sniffing digits to keep it as efficient as possible. Please move back to the previous version.

File: lib/image-diff.js [hunk 2]
```diff
@@ -81,8 +81,8 @@ ImageDiff.createDiff = function (options, cb) {
     }
 
     // Callback with pass/fail
-    var totalDifference = resultInfo[1];
-    return cb(null, totalDifference === '0');
+    var totalDifference = resultInfo[2]; // this is the percentage difference
+    return cb(null, parseFloat(totalDifference) <= options.threshold, options.diffPath);
```
Comment (twolfson) [inline lib/image-diff.js]: Also, we will need to fallback `options.threshold` to `0`.

Comment (twolfson) [inline lib/image-diff.js]: Oh, and we need to specify the radix for `parseFloat` to avoid parsing it as an octal.

```js
parseFloat(totalDifference, 10);
```

File: lib/image-diff.js [hunk 3]
```diff
@@ -81,8 +81,8 @@ ImageDiff.createDiff = function (options, cb) {
     }
 
     // Callback with pass/fail
-    var totalDifference = resultInfo[1];
-    return cb(null, totalDifference === '0');
+    var totalDifference = resultInfo[2]; // this is the percentage difference
```
Comment (twolfson) [inline lib/image-diff.js]: We should be renaming this variable since it's now the percentage difference, not the total difference.

File: lib/image-diff.js [hunk 4]
```diff
@@ -202,7 +205,7 @@ ImageDiff.prototype = {
         fs.unlink(filepath, cb);
       }, function callOriginalCallback (_err) {
         // Callback with the imagesAreSame
-        callback(err, imagesAreSame);
+        callback(err, imagesAreSame, diffPath);
```
Comment (twolfson) [inline lib/image-diff.js]: To keep the API as simple as possible for future changes, please don't send back the `diffPath`. The user already knows what the is since they gave it to us.

**Comment by twolfson** [comment]
The gist is here but we need tests. Can you add some?

File: lib/image-diff.js (hunk 2 above)
Comment (twolfson) [inline lib/image-diff.js]: And as mentioned in the originally proposed changes, we need to handle when this parses to `NaN`:

https://github.com/uber/image-diff/issues/14#issuecomment-86199714

```js
if (isNaN(percentDifference)) {
  return cb(new Error('Attempted to parse percentage difference from `image-diff\'s stderr` but got `NaN` from "' + resultInfo[2] + '"'));
}
```

**Comment by nikchosh** [comment]
I have included all your suggestions (ultra new to git -- sorry to make you my mentor :) ) I need to look at the tests. I am tester, and getting feet wet in Node development

File: lib/image-diff.js (hunk 1 above)
Comment (nikchosh) [inline lib/image-diff.js]: I remember why I changed this - in case of only a few pixel difference, we get a result like this from compare.exe:
  Channel distortion: RMSE
    red: 0.00129297 (5.07046e-006)
    green: 0 (0)
    blue: 0.00158356 (6.21002e-006)
    all: 0.00118031 (4.62868e-006)
So, we can either use \(.*\) or /\d+\.?\d*(e-\d+)?/ to allow for something small

Comment (twolfson) [inline lib/image-diff.js]: Ah, interesting. We should probably isolate the parser to a separate function so we can test these individually. I imagine it will become very tedious to create a set of separate images for each of these cases.

As for the regexp, being explicit is always more useful than being implicit. Additionally, we should probably include examples of the different values in comments. The regexp should look like:

```js
/all: (\d+\.?\d*) \((\d+\.?\d*e?\-?\d*)\)/)
```

This allows for `0`, `0.1`, `0.1e-3`, and even unexpected cases like `1e3`, `1.0e3`.

**Comment by twolfson** [comment]
I can't see your changes. Please push them to update the PR (pull request).

File: lib/image-diff.js [hunk 5]
```diff
@@ -81,8 +81,12 @@ ImageDiff.createDiff = function (options, cb) {
     }
 
     // Callback with pass/fail
-    var totalDifference = resultInfo[1];
-    return cb(null, totalDifference === '0');
+    var percentDifference = parseFloat(resultInfo[2], 10);
+    var threshold = (typeof(options.threshold) === 'undefined' || isNaN(parseFloat(options.threshold, 10))) ? 0.0 : parseFloat(options.threshold);
```
Comment (twolfson) [inline lib/image-diff.js]: There is a lot going on in this line.

- Using `typeof` to detect undefined is only necessary when looking at a global (e.g. `typeof window === 'undefined')
    - Otherwise, we can use a simpler direct comparison with `undefined` (e.g. `options.foo === undefined`)
- We don't need to detect `undefined`, a fallback will work fine (e.g. `var foo = options.bar || baz;`)
- We don't need to be defensive and coerce `options.threshold`. We can assume the developer wants the library to work and will pass in a `Number` (i.e. no `isNaN/parseFloat` necessary)
- In JS, all `Numbers` are 64 bit floats. The decimal isn't necessary; `0` will work fine.
    - https://developer.mozilla.org/en-US/docs/Glossary/Number

The simplified result should be:

```js
var threshold = options.threshold || 0;
```

**Comment by twolfson** [comment]
Getting closer, still needs tests though =/

**Comment by nikchosh** [comment]
Tests added - probably documentation needs to be updated too

File: test/image-diff_test.js [hunk 6]
```diff
@@ -31,7 +31,7 @@ describe('image-diff', function () {
     runImageDiff({
       actualImage: __dirname + '/test-files/checkerboard.png',
       expectedImage: __dirname + '/test-files/white.png',
-      diffImage: __dirname + '/actual-files/different.png'
+      diffImage: __dirname + '/actual-files/different.png',
```
Comment (twolfson) [inline test/image-diff_test.js]: Can you move back this line? It is noise in the PR =/

File: test/image-diff_test.js [hunk 7]
```diff
@@ -45,12 +45,51 @@ describe('image-diff', function () {
     });
   });
 
+  describe('diffing different images using lower threshold', function () {
+    runImageDiff({
+      actualImage: __dirname + '/test-files/checkerboard.png',
+      expectedImage: __dirname + '/test-files/white.png',
+      diffImage: __dirname + '/actual-files/different.png',
```
Comment (twolfson) [inline test/image-diff_test.js]: `different` is very ambiguous. Can we name the file `threshold-under` so it is semantic to the test?

File: test/image-diff_test.js [hunk 8]
```diff
@@ -45,12 +45,51 @@ describe('image-diff', function () {
     });
   });
 
+  describe('diffing different images using lower threshold', function () {
```
Comment (twolfson) [inline test/image-diff_test.js]: This doesn't read that well. Can we update it to:

```js
comparing different images under a threshold
```

File: test/image-diff_test.js [hunk 9]
```diff
@@ -45,12 +45,51 @@ describe('image-diff', function () {
...
+      actualImage: __dirname + '/test-files/checkerboard.png',
+      expectedImage: __dirname + '/test-files/white.png',
+      diffImage: __dirname + '/actual-files/different.png',
+      threshold: 0.7
+    });
+    imageUtils.loadActual('different.png');
+    imageUtils.loadExpected('different.png');
+
+    it('asserts images are different', function () {
+      assert.strictEqual(this.imagesAreSame, false);
```
Comment (twolfson) [inline test/image-diff_test.js]: Don't we want this to be asserting the images are the same?

```
it('asserts images are the similar enough', function () {
  assert.strictEqual(this.imagesAreSame, true);
});

Comment (twolfson) [inline test/image-diff_test.js]: Ah, I think I confused these 2 text cases. Ignore this comment =/

File: test/image-diff_test.js (hunk 8 above)
Comment (twolfson) [inline test/image-diff_test.js]: Due to my confusion, I had this test case wrong. It should be named something like:

```
comparing different images over a threshold
```

File: test/image-diff_test.js [hunk 10]
```diff
@@ -45,12 +45,51 @@ describe('image-diff', function () {
     });
   });
 
+  describe('diffing different images using lower threshold', function () {
+    runImageDiff({
+      actualImage: __dirname + '/test-files/checkerboard.png',
+      expectedImage: __dirname + '/test-files/white.png',
+      diffImage: __dirname + '/actual-files/different.png',
+      threshold: 0.7
```
Comment (twolfson) [inline test/image-diff_test.js]: Can we make this something like `0.2` to make it a very obvious threshold difference to be over?

File: test/image-diff_test.js [hunk 11]
```diff
@@ -45,12 +45,51 @@ describe('image-diff', function () {
...
+    it('asserts images are different', function () {
+      assert.strictEqual(this.imagesAreSame, false);
+    });
+
+    it('writes a highlighted image diff to disk', function () {
+      assert.deepEqual(this.actualPixels, this.expectedPixels);
+    });
+  });
+
+  describe('diffing different images using higher threshold', function () {
```
Comment (twolfson) [inline test/image-diff_test.js]: As with the previous test case, we should update the description to make it more clear:

```
comparing different images over a threshold
```

File: test/image-diff_test.js [hunk 12]
```diff
@@ -45,12 +45,51 @@ describe('image-diff', function () {
...
+    runImageDiff({
+      actualImage: __dirname + '/test-files/checkerboard.png',
+      expectedImage: __dirname + '/test-files/white.png',
+      diffImage: __dirname + '/actual-files/different2.png',
+      threshold: 0.8
```
Comment (twolfson) [inline test/image-diff_test.js]: Can we make this `0.9` to make it *really* obvious

File: test/image-diff_test.js [hunk 13]
```diff
@@ -45,12 +45,51 @@ describe('image-diff', function () {
...
+    it('writes a highlighted image diff to disk', function () {
+      assert.deepEqual(this.actualPixels, this.expectedPixels);
+    });
+  });
+
+  describe('diffing different images using higher threshold', function () {
+    runImageDiff({
+      actualImage: __dirname + '/test-files/checkerboard.png',
+      expectedImage: __dirname + '/test-files/white.png',
+      diffImage: __dirname + '/actual-files/different2.png',
```
Comment (twolfson) [inline test/image-diff_test.js]: As with the other comment, can we rename this to be something like `threshold-under.png`?

File: test/image-diff_test.js (hunk 11 above)
Comment (twolfson) [inline test/image-diff_test.js]: Ah, I meant under. This stuff is confusing x_x

```
comparing different images under a threshold
```

File: test/image-diff_test.js [hunk 14]
```diff
@@ -45,12 +45,51 @@ describe('image-diff', function () {
...
+    imageUtils.loadExpected('different.png');
+
+    it('asserts images are different', function () {
+      assert.strictEqual(this.imagesAreSame, true);
+    });
+
+    it('writes a highlighted image diff to disk', function () {
+      assert.deepEqual(this.actualPixels, this.expectedPixels);
+    });
+  });
```
Comment (twolfson) [inline test/image-diff_test.js]: These are more edge case tests and I would prefer to run them later on. Can we relocate them after `diffing the same image`?

File: test/image-diff_test.js [hunk 15]
```diff
@@ -45,12 +45,51 @@ describe('image-diff', function () {
...
+
   describe('diffing the same image', function () {
     runImageDiff({
       actualImage: __dirname + '/test-files/checkerboard.png',
       expectedImage: __dirname + '/test-files/checkerboard.png',
       diffImage: __dirname + '/actual-files/same.png'
     });
+
```
Comment (twolfson) [inline test/image-diff_test.js]: Please remove this change. It's PR noise.

File: test/image-diff_test.js [hunk 16]
```diff
@@ -67,7 +106,8 @@ describe('image-diff', function () {
     runImageDiff({
       actualImage: __dirname + '/test-files/checkerboard-excess.png',
       expectedImage: __dirname + '/test-files/checkerboard.png',
-      diffImage: __dirname + '/actual-files/different-size.png'
+      diffImage: __dirname + '/actual-files/different-size.png',
+      threshold: 0.70
```
Comment (twolfson) [inline test/image-diff_test.js]: We should not need to update this test for threshold or similar. Please change it back.

**Comment by twolfson** [comment]
Everything looks fine. Mostly fine tuning things now.

**Comment by nikchosh** [comment]
Taken all your comments into account - please review

**Comment by twolfson** [comment]
There seem to still be a few lingering inline comments that I would like to get sorted out. Can we take care of those?

**Comment by nikchosh** [comment]
Would you like me to remove all // comments DEV: etc.

**Comment by twolfson** [comment]
Nope, I am referring to these:

https://github.com/uber/image-diff/pull/15/files#r27454478

https://github.com/uber/image-diff/pull/15/files#r27454735

https://github.com/uber/image-diff/pull/15/files#r27454747

**Comment by nikchosh** [comment]
Hi Todd, any way to merge the changes? :)

**Comment by twolfson** [comment]
There is still this trailing comma comment left =/

https://github.com/uber/image-diff/pull/15/files#r27454478

Additionally, we seem to be blocked by #16. I will bump the issue so it can be addressed.

**Comment by nikchosh** [comment]
Commas fixed

**Comment by twolfson** [comment]
Sweet, everything looks great now. Still waiting on a reply in #16

**Comment by mlmorg** [comment]
Can we rebase and squash into a single commit? Thanks!

**Comment by twolfson** [comment]
This PR hasn't been active in over 1 year. We can probably close it due to inactivity
